core/plugin_manager.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Manager for all items registered with the workspace.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+
+'use strict';
+
+goog.provide('Blockly.PluginManager');
+
+goog.require('Blockly.IPlugin');
+
+
+/**
+ * Manager for all items registered with the workspace.
+ * @constructor
+ */
+Blockly.PluginManager = function() {
+  /**
+   * A map of the plugins registered with the workspace, mapped to id.
+   * @type {!Object<string, !Blockly.PluginManager.PluginData>}
+   */
+  this.pluginData_ = {};
+
+  /**
+   * A map of types to plugin ids.
+   * @type {!Object<string, Array<string>>}
+   */
+  this.typeToPluginId_ = {};
+};
+
+/**
+ * @typedef {{
+ *            id: string,
+ *            plugin: !Blockly.IPlugin,
+ *            types: !Array<typeof Blockly.IPlugin>,
+ *            weight: number
+ *          }}
+ */
+Blockly.PluginManager.PluginData;
+
+/**
+ * Adds a plugin.
+ * @param {!Blockly.PluginManager.PluginData} pluginDataObject The plugin.
+ */
+Blockly.PluginManager.prototype.addPlugin = function(pluginDataObject) {
+  this.pluginData_[pluginDataObject.id] = pluginDataObject;
+  for (var i = 0, type; (type = pluginDataObject.types[i]); i++) {
+    var typeKey = type.name;  // The map key needs to be a string
```

Comment:
I would be careful with using Function.name. It's not supported in IE 11:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name

and it's also not consistent across browsers, e.g: I tried this on my machine, and it's an empty string for `Blockly.IPositionable.name`

Reply:
code updated